### PR TITLE
Processing Data in Models

### PR DESCRIPTION
**Pull request by tan12082001**
# [Blog App - processing data in models](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/Processing_data_model_project_v1_1.md#blog-app---processing-data-in-models)

### Implemented Changes:
* Created the Model classes `User, Post, Comment and Like` inherited from the `ApplicationRecord` Model.
* Updated the model classes with associations such as `has_many, belongs_to`.
* Included the foreign_key for the Post model with author_id which will be added in the User class.
* Using the models, added data to Users, posts to a user, comments to a user.
* Implemented simple CRUD methods.
* Updated the Class methods `Post, Comment, and Like` with the custom counter method that updates the counter attribute using the `after_save` callback.
* Updated the `User and Post` classes with `three_recent_posts and five_recent_comments` methods.
#### Create, Read
image
image
#### Update
image
#### Delete
image

### CODE REVIEW

File: app/models/post.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+class Post < ApplicationRecord
+  belongs_to :author, class_name: 'User'
+  has_many :comment
+  has_many :like
+  after_save :update_posts_counter
+
```
Comment (prolajumokeoni) [inline app/models/post.rb:6]: - [x] Please make your `update_posts_counter `method to be set to` private`, which will make your class to be called only within the Post class 👍🏿

File: app/models/post.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
+class Post < ApplicationRecord
+  belongs_to :author, class_name: 'User'
+  has_many :comment
+  has_many :like
```
Comment (prolajumokeoni) [inline app/models/post.rb]: - [x] Kindly update your post associations  to 👍🏿 
  has_many :comments
  has_many :likes
```

File: app/models/user.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+class User < ApplicationRecord
+  has_many :post, foreign_key: 'author_id'
+  has_many :comment
+  has_many :like
```
Comment (prolajumokeoni) [inline app/models/user.rb]: - [x] Kindly update your user associations  to
 ```
 has_many :posts, foreign_key: 'author_id'
  has_many :comments
  has_many :likes
```

File: app/models/comment.rb [hunk 4]
```diff
@@ -0,0 +1,10 @@
+class Comment < ApplicationRecord
+  belongs_to :post
+  belongs_to :user
+
+  after_save :update_comments_counter
```
Comment (prolajumokeoni) [inline app/models/comment.rb:5]: - [x] Please make your `update_posts_counter `method to be set to` private`, which will make your class to be called only within the comment class 👍🏿

File: app/models/post.rb [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+  after_save :update_posts_counter
+
+  def update_posts_counter
+    author.increment!(:posts_counter)
+  end
+
+  def five_recent_comments
+    Comment.where(post_id: id).order(created_at: :desc).limit(5)
```
Comment (prolajumokeoni) [inline app/models/post.rb]: - [x] In the method `five_recent_comments`  it should be `comment` not Comment as you are targeting to get the five recent comments 👍🏿

File: app/models/like.rb [hunk 6]
```diff
@@ -0,0 +1,10 @@
+class Like < ApplicationRecord
+  belongs_to :post
+  belongs_to :user
+
+  after_save :update_likes_counter
```
Comment (prolajumokeoni) [inline app/models/like.rb:5]: - [x]  Please make your update_likes_counter method to be set to private, which will make your class to be called only within the like class 👍🏿

File: app/models/user.rb [hunk 7]
```diff
@@ -0,0 +1,9 @@
...
+
+  def three_recent_posts
+    Post.where(author: self).order(created_at: :desc).limit(3)
```
Comment (prolajumokeoni) [inline app/models/user.rb]: In the method three_recent_poss tit should be posts not Post as you are targeting to get the five recent comments 👍🏿

**Review by prolajumokeoni — CHANGES_REQUESTED** [review]
Hi @tan12082001 
Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
- Added User, Post, Comment and Like models inside app/models/ directory ✅



## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- N/A


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification @prolajumokeoni.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by Felix45 — APPROVED** [review]
# Approved :green_circle: 
Hi @tan12082001, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



## Highlights :heavy_check_mark: 
- [x] All models have been set correctly :white_check_mark: 
- [x] All migrations run as expected :white_check_mark: 
- [x] Application works as expected when using rails console :white_check_mark:  
- [x] Good descriptive commit messages :white_check_mark:  
- [x] Pull request has a descriptive title and summary :white_check_mark: 
- [x] Repository has a professional readme file :white_check_mark:  

## Optional suggestions
- [ Optional ] I think it would be nice if you remove the `Live demo` section from your readme because this project has not been deployed online yet. The live demo section on your readme is a bit misleading because there's no live link available at this time. I highly recommend you rectify this in order to make the documentation of your repository professional.

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**